File: easy/14_Longest_Common_Prefix.py
# Write a function to find the longest common prefix string amongst an array of strings.
#
# If there is no common prefix, return an empty string ""

class Solution:
    def longestCommonPrefix(self, strs: list[str]) -> str:
        tmp = strs[0]
        strs.sort(key=len)
        result = ''
        for word in strs:
            result = ''
            for sym_pref, sym_el in zip(tmp, word):
                if sym_el == sym_pref:
                    result += sym_el
                else:
                    break
            tmp = result
        return result


a = Solution()
print(a.longestCommonPrefix(["aa", "ab"]))

File: go/easy/14_longest_common_prefix.go
package main

import (
	"fmt"
	"strings"
)

//
//Write a function to find the longest common prefix string amongst an array of strings.
//
//If there is no common prefix, return an empty string "".
//
//Example 1:
//
//Input: strs = ["flower","flow","flight"]
//Output: "fl"
//Example 2:
//
//Input: strs = ["dog","racecar","car"]
//Output: ""
//Explanation: There is no common prefix among the input strings.

func longestCommonPrefix(strs []string) string {
	if len(strs) == 0 {
		return ""
	}

	var res strings.Builder
	minLen := len(strs[0])

	for _, str := range strs {
		if len(str) < minLen {
			minLen = len(str)
		}
	}

	for i := 0; i < minLen; i++ {
		curEl := strs[0][i : i+1]

		for _, str := range strs {
			if str[i:i+1] != curEl {
				return res.String()
			}
		}

		res.WriteString(curEl)
	}

	return res.String()
}

func main() {
	fmt.Println(longestCommonPrefix([]string{"dog", "racecar", "car"}))
	fmt.Println(longestCommonPrefix([]string{"flower", "flow", "flight"}))
}
